### PR TITLE
polys: make ANP use DMP internally

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Followup from gh-25722

#### Brief description of what is fixed or changed

Make the ANP class that is used for algebraic numberfields use DMP internally. This means that ANP can automatically benefit from any speedups in DMP that arise from using e.g. Flint rather than the existing `dmp_*` functions.

#### Other comments

The same should also be done for `DMF`.

#### Release Notes

(I will add a release note later)

* polys
    * The ANP class which is used to represent elements of an algebraic field is now built on top of the DMP class which is used to represent dense polynomials. This means that algebraic number fields like `QQ<sqrt(2)>` can benefit from the speed improvements of python-flint when the ground types are set to flint.

### CODE REVIEW

File: sympy/polys/polyclasses.py [hunk 1]
```diff
@@ -2759,35 +2759,49 @@ def init_normal_ANP(rep, mod, dom):
...
 
+    @property
+    def rep(self):
+        raise RuntimeError("ANP.rep")
+        return self._rep.to_list()
+
+    @property
+    def mod(self):
+        raise RuntimeError("ANP.mod")
+        return self._mod.to_list()
```
Comment (sylee957) [inline sympy/polys/polyclasses.py:2798]: This always raise errors. Are you leaving this because this is part of some incomplete work?

Comment (oscarbenjamin) [reply sympy/polys/polyclasses.py:2798]: It is incomplete. I haven't decided whether or not to deprecate these. The `rep` attribute can be replaced with `.to_list()` just like for `DMP`. The mod argument does not have an existing replacement. It is perhaps also not too important to deprecate them but I do want to make sure that they aren't use within the sympy codebase itself.

Comment (sylee957) [reply sympy/polys/polyclasses.py]: I think that although keeping the `rep, dom` as list is backward compatible, I don't think that it is more useful than using instance of `DMP` directly.

Comment (oscarbenjamin) [reply sympy/polys/polyclasses.py]: I guess at least there should be new public methods/attributes to get the rep and mod as DMP.

**Comment by oscarbenjamin** [comment]
Current state:
```console
$ cat t.py 
from sympy import *

N = 20

M = randMatrix(N) + randMatrix(N)*sqrt(2) + randMatrix(N)*sqrt(3) + randMatrix(N)*sqrt(5)

dM = M.to_DM(extension=True)

dM.inv()

$ time python t.py

real	0m38.798s
user	0m38.550s
sys	0m0.214s

$ time SYMPY_GROUND_TYPES=flint python t.py

real	0m2.179s
user	0m2.013s
sys	0m0.170s
```
More can be done to improve that but maybe for now this is good enough in terms of basic speed. Profiles suggest that most of the time is spent in python-flint rather than in overheads in the Python code so big speed improvements would need to be from a change of algorithm rather than micro-optimisation (at least for the things I have tested).

**Comment by oscarbenjamin** [comment]
I've optimised some of the ANP methods a bit more.

This is without flint:
```python
In [2]: cat t.py
from sympy import *

from time import time

def rand(N):
    M = (
        randMatrix(N)
        + randMatrix(N)*sqrt(2)
        + randMatrix(N)*sqrt(3)
        + randMatrix(N)*sqrt(5)
        + randMatrix(N)*sqrt(7)
        + randMatrix(N)*sqrt(11)
    )
    return M.to_DM(extension=True)

In [3]: run t.py

In [4]: for n in range(1, 5):
   ...:     dM = rand(n)
   ...:     start = time()
   ...:     dM.inv()
   ...:     time_taken = time() - start
   ...:     print(f'{n}x{n} matrix: {time_taken:.3g} seconds')
   ...: 
1x1 matrix: 0.193 seconds
2x2 matrix: 2.39 seconds
3x3 matrix: 6.87 seconds
4x4 matrix: 18.1 seconds
```
This is with flint:
```python
In [3]: for n in range(1, 10):
   ...:     dM = rand(n)
   ...:     start = time()
   ...:     dM.inv()
   ...:     time_taken = time() - start
   ...:     print(f'{n}x{n} matrix: {time_taken:.3g} seconds')
   ...: 
1x1 matrix: 0.00954 seconds
2x2 matrix: 0.0395 seconds
3x3 matrix: 0.106 seconds
4x4 matrix: 0.218 seconds
5x5 matrix: 0.406 seconds
6x6 matrix: 0.679 seconds
7x7 matrix: 1.11 seconds
8x8 matrix: 1.61 seconds
9x9 matrix: 2.27 seconds
```
This is about 100x faster for a 4x4 matrix but the difference increases for larger sizes.

**Comment by oscarbenjamin** [comment]
I noticed while timing this that primitive_element still internally uses `dup_zz_factor` rather than using flint for the factorisation. This is because it factorises polynomials over e.g. `QQ<sqrt(2)>` which will follow the existing `dmp_*` functions and at the bottom they end up calling `dup_zz_factor`. It should be changed so that some lower level functions like `dup_zz_factor` call out to Flint.

**Comment by oscarbenjamin** [comment]
I think that the main thing missing here is just to add public methods for getting `rep` and `mod` as `DMP` rather than as lists.

**Comment by oscarbenjamin** [comment]
This is now ready fir reivew

**Comment by jksuom** [comment]
Looks good.

**Comment by oscarbenjamin** [comment]
Thanks!

**Pull request merged**